linalg/AmgXSolver.hpp

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//Reference:
+//Pi-Yueh Chuang, & Lorena A. Barba (2017).
+//AmgXWrapper: An interface between PETSc and the NVIDIA AmgX library. J. Open Source Software, 2(16):280, doi:10.21105/joss.00280
+
+/** \brief A macro to check the returned CUDA error code.
+ *
+ * \param call [in] Function call to CUDA API.
+ */
+
+ #ifndef MFEM_AMGX_SOLVER
+ #define MFEM_AMGX_SOLVER
+
+ #include "../config/config.hpp"
+
+ #ifdef MFEM_USE_AMGX
+ #ifdef MFEM_USE_MPI
+
+ #include <amgx_c.h>
+ #include <mpi.h>
+ #include "hypre.hpp"
+
+# define CHECK(call)                                                        \
+  {                                                                           \
+  const cudaError_t       error = call;                                   \
+  if (error != cudaSuccess)                                               \
+    {                                                                       \
+  printf("Error: %s:%d, code:%d, reason: %s\n",                         \
+         __FILE__, __LINE__, error, cudaGetErrorString(error));          \
+    }                                                                       \
+  }
+
+
+namespace mfem
+{
+
+class AmgXSolver : public Solver
+{
+
+public:
+
+
+  AmgXSolver() = default;
+
+  AmgXSolver(const MPI_Comm &comm,
+             const std::string &modeStr, const std::string &cfgFile, int &nDevs);
+
+  AmgXSolver(const std::string &modeStr, const std::string &cfgFile);
+
+  ~AmgXSolver();
+
+  void initialize(const MPI_Comm &comm,
+                  const std::string &modeStr, const std::string &cfgFile, int &nDevs);
+
+  void initialize(const std::string &modeStr, const std::string &cfgFile);
+
+  void finalize();
+
+  void setA(const mfem::HypreParMatrix &A);
+
+  virtual void SetOperator(const Operator &op);
+
+  virtual void SetOperator(const SparseMatrix &in_A);
```

Comment:
Here you should be able to only have 1 set operator method that takes in an Operator type. Then depending on what is passed in do the right thing (there should be a way to check what type of object is passed in).